store/tikv/region_cache.go

```diff
@@ -66,32 +79,74 @@ func (c *RPCContext) GetStoreID() uint64 {
 	return 0
 }
 
-// GetRPCContext returns RPCContext for a region. If it returns nil, the region
-// must be out of date and already dropped from cache.
-func (c *RegionCache) GetRPCContext(bo *Backoffer, id RegionVerID) (*RPCContext, error) {
-	c.mu.RLock()
-	region, ok := c.mu.regions[id]
+// DropInvalidRegion returns true if the region is invalid and then dropped or if it has been dropped
+func (c *RegionCache) DropInvalidRegion(id RegionVerID) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cachedregion, ok := c.mu.regions[id]
 	if !ok {
-		c.mu.RUnlock()
-		return nil, nil
-	}
-	kvCtx := region.GetContext()
-	c.mu.RUnlock()
-
-	addr, err := c.GetStoreAddr(bo, kvCtx.GetPeer().GetStoreId())
-	if err != nil {
-		return nil, errors.Trace(err)
+		return true
 	}
-	if addr == "" {
-		// Store not found, region must be out of date.
-		c.DropRegion(id)
-		return nil, nil
+	// we need to check ttl because concurrent threads may fresh it
+	// For example, T1 find ttl valid but before T1 freshes the ttl, T2 find ttl invalid
+	// In this case, T2 will invoke this function but T1 may have freshed the ttl
+	// Then, T2 should not drop the cached region
+	ttl := atomic.LoadInt64(&cachedregion.ttl) // we may not need an atomic load, I think a simple load is enough?
```

Comment:
I think we don't need to check the access time again, just delete it.

Reply:
Indeed, both rechecking and directly deleting are correct.
Directly deleting may simplify the code.
However, I think it leads to strange cases. Suppose that there are two visits on the same cached region. If the first visit determines to refresh the ttl of the region,  when it conducts this action, it may find that the region was removed because of out-of-date. This is strange.

Reply:
The first reader found the region is valid, then he will just refersh the region, he won't lookup the map for the region again, so he never find that region was removed.

Reply:
Yes，You are right.
Either deleting or rechecking is correct.
The choice depends on whether we want the execution is equivalent to the actual happening order of the two operations.
By rechecking, I choose that the execution should be equivalent to the actual happening order of the two operations.
This may be more strict than directly deleting. 
I think the two possible choices are OK.